src/Columns/ColumnVector.cpp

```diff
@@ -337,6 +371,47 @@ void ColumnVector<T>::updatePermutation(IColumn::PermutationSortDirection direct
             bool use_radix_sort = (sort_is_stable && ascending && !std::is_floating_point_v<T>) || !sort_is_stable;
             size_t size = end - begin;
 
+            /// A case for counting sort
+            if constexpr (std::is_same_v<T, UInt8>)
+            {
+                if (end - begin > 256)
+                {
+                    size_t cnt[256];
+                    memset(cnt, 0, sizeof(cnt));
+
+                    for (auto * it = begin; it != end; ++it)
+                        cnt[data[*it]]++;
```

Comment:
maybe we should write here at `data[*it] +- 1` to avoid shifting array by one below.